homeassistant/components/numato/binary_sensor.py

```diff
@@ -0,0 +1,98 @@
+"""
+Binary sensor platform integration for Numato USB GPIO expanders.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/integrations/numato#binary-sensor
+"""
+import logging
+
+from numato_gpio import NumatoGpioError
+
+from homeassistant.components.binary_sensor import BinarySensorDevice
+from homeassistant.const import DEVICE_DEFAULT_NAME
+
+from . import CONF_BINARY_SENSORS, CONF_ID, DOMAIN, edge_detect, read_input, setup_input
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_INVERT_LOGIC = "invert_logic"
+CONF_PORTS = "ports"
+CONF_DEVICES = "devices"
+CONF_DEVICE_ID = "id"
+DEFAULT_INVERT_LOGIC = False
+
+DEPENDENCIES = ["numato"]
+
+
+# pylint: disable=unused-argument
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the configured Numato USB GPIO binary sensor ports."""
+
+    binary_sensors = []
+    devices = hass.data[DOMAIN]
+    for device in [d for d in devices if CONF_BINARY_SENSORS in d]:
+        device_id = device[CONF_ID]
+        platform = device[CONF_BINARY_SENSORS]
+        invert_logic = platform[CONF_INVERT_LOGIC]
+        ports = platform[CONF_PORTS]
+        for port_id, port_name in ports.items():
+            try:
+                binary_sensors.append(
+                    NumatoGPIOBinarySensor(port_name, device_id, port_id, invert_logic)
+                )
+            except NumatoGpioError as ex:
+                _LOGGER.error(
+                    "Numato USB device %s port %s failed: %s",
+                    device_id,
+                    port_id,
+                    str(ex),
+                )
+    add_devices(binary_sensors, True)
+
+
+class NumatoGPIOBinarySensor(BinarySensorDevice):
+    """Represents a binary sensor (input) port of a Numato GPIO expander."""
+
+    def __init__(self, name, device_id, port, invert_logic):
+        """Initialize the Numato GPIO based binary sensor."""
+        # pylint: disable=no-member
+        self._name = name or DEVICE_DEFAULT_NAME
+        self._device_id = device_id
+        self._port = port
+        self._invert_logic = invert_logic
+        self._state = None
+        setup_input(self._device_id, self._port)
```

Comment:
We don't want side effects in init method. Either do this before creating the entity or in the entity coroutine method `async_added_to_hass`.

Reply:
If I do understand async right, I shouldn't use sync blocking API (which numato-gpio is) in an async method like an `async_added_to_hass` correct? If so, this would leave option a) before creating the entity.
Please correct me if I'm wrong.

Reply:
We can schedule sync work on the executor thread pool from coroutines, with `hass.async_add_executor_job`.

Reply:
Done.